### PR TITLE
[Zerobubble] merge main.

### PR DESCRIPTION
**Pull request by duanjunwen**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: .github/workflows/build_on_pr.yml [hunk 1]
```diff
@@ -140,7 +140,7 @@ jobs:
 
       - name: Install Colossal-AI
         run: |
-          BUILD_EXT=1 pip install -v -e .
+          BUILD_EXT=1 pip install -v .
```
Comment (TongLi3701) [inline .github/workflows/build_on_pr.yml:143]: We do not need -e on PRs

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 2]
```diff
@@ -0,0 +1,965 @@
...
+
+        if self.forward_only:
+            self.num_microbatch = (self.batch_size - 1) // self.microbatch_size + 1
+            # NOTE: disable metadata cache when batch size changes (not valid anymore)
+            # if self.batch_size != self.last_batch_size:
+            # self.enable_metadata_cache = False
+            # self.send_tensor_metadata = True
+            # self.send_grad_metadata = True
+            # self.tensor_metadata_recv = None
+            # self.grad_metadata_recv = None
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py:195]: > why

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6142/commits/9aa4c67a6351ec550de1b60125e09fc9b07ef194

File: colossalai/pipeline/schedule/v_schedule.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+                    if fill_f:
+                        put_w(i)
+                put(0, put_item, i)
+
+        for i in range(self.n_stage):
+            while len(pending_w[i]) > 0:
+                put_w(i)
+
+        # for i in range(self.n_stage):
+        #     print(stage_str[i])
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/pipeline/schedule/v_schedule.py]: > remove such debug info

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6142/commits/9aa4c67a6351ec550de1b60125e09fc9b07ef194

File: colossalai/pipeline/schedule/v_schedule.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+                        put(0, iter_chunk_, j)
+                iter_chunk_ = 1 - iter_chunk_
+            # while mem[i] + self.fbw_mem[0] <= self.max_mem and cur_time[i] + self.fbw_cost[0] <= tmp:
+            #     if iter_chunk_ == 0 and count[i][0] >= count[i - 1][0]:
+            #         break
+            #     for j in range(self.n_stage - 1, i - 1, -1):
+            #         if count[j][iter_chunk_] < self.n_micro:
+            #             put(0, iter_chunk_, j)
+            #     iter_chunk_ = 1 - iter_chunk_
+            # end_tmp = max(tmp, cur_time[i]) + self.fbw_cost[1]
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply colossalai/pipeline/schedule/v_schedule.py]: > Is this useless?

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6142/commits/9aa4c67a6351ec550de1b60125e09fc9b07ef194

File: .github/workflows/build_on_schedule.yml [hunk 5]
```diff
@@ -55,7 +55,7 @@ jobs:
         if: steps.check-avai.outputs.avai == 'true'
         run: |
           [ ! -z "$(ls -A /github/home/cuda_ext_cache/)" ] && cp -r /github/home/cuda_ext_cache/* /__w/ColossalAI/ColossalAI/
-          BUILD_EXT=1 pip install -v -e .
+          BUILD_EXT=1 pip install -v .
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply .github/workflows/build_on_schedule.yml]: > keep this

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6142/commits/9aa4c67a6351ec550de1b60125e09fc9b07ef194

File: .github/workflows/build_on_pr.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (duanjunwen) [reply .github/workflows/build_on_pr.yml]: > keep this

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6142/commits/9aa4c67a6351ec550de1b60125e09fc9b07ef194

**Review by ver217 — APPROVED** [review]

**Pull request merged**